src/OpenTracing/Util/AsyncLocalScope.cs

```diff
@@ -0,0 +1,43 @@
+namespace OpenTracing.Util
+{
+    /// <summary>
+    /// The <see cref="AsyncLocalScope"/> is a simple <see cref="IScope"/> implementation
+    /// that relies on C#'s AsyncLocal/CallContext storage primitive.
+    /// </summary>
+    /// <seealso cref="IScopeManager"/>
+    public class AsyncLocalScope : IScope
+    {
+        private readonly AsyncLocalScopeManager _scopeManager;
+        private readonly ISpan _wrappedSpan;
+        private readonly bool _finishOnDispose;
+        private readonly IScope _scopeToRestore;
+
+        public AsyncLocalScope(AsyncLocalScopeManager scopeManager, ISpan wrappedSpan, bool finishOnDispose)
+        {
+            _scopeManager = scopeManager;
+            _wrappedSpan = wrappedSpan;
+            _finishOnDispose = finishOnDispose;
+
+            _scopeToRestore = scopeManager.Active;
+            scopeManager.Active = this;
+        }
+
+        public ISpan Span => _wrappedSpan;
+
+        public void Dispose()
+        {
+            if (_scopeManager.Active != this)
+            {
+                // This shouldn't happen if users call methods in the expected order. Bail out.
```

Comment:
Throw `InvalidOperationException`?

```
The exception that is thrown when a method call is invalid for the object's current state
```

Reply:
Depends on your views... Throw from Dispose is generally frowned upon, but which is worse:
1) Failing silently leaving a bug to go potentially unnoticed for quite some time and hard to debug
2) Throwing from Dispose on code errors

The reason Throw on Dispose is frowned upon is because the following
```C#
try
{
 using (var span = /* ... */)
 {
   // Do stuff
   throw new IOException();
 }
}
catch (IOException)
{
 // Not hit if (non-IOException) thrown from Dispose, and the exception being thrown is not available to `Dispose` so it can't even be wrapped
}
```

But I generally feel that's okay because you only loose an exception if you're 1) already in an exceptional state and 2) have done something exceptional with the library that suggests a developer bug (I avoid it when it's a state-based/concurrency-based bug but when it's something that's likely just a coding error it seems better to fail-fast to me)

Reply:
AFAIK not throwing an exception here has been a deliberate decision in Java as tracers are not meant to break the application (in the same way that loggers are not meant to break it). So if you want to have this changed we should discuss it in a separate issue.